src/plugins/intel_gpu/src/plugin/transformations/move_convert_after_gather.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "move_convert_after_gather.hpp"
+
+#include "openvino/core/rt_info.hpp"
+#include "openvino/op/constant.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "transformations/utils/utils.hpp"
+#include "transformations/rt_info/decompression.hpp"
+
+namespace ov {
+namespace intel_gpu {
+
+MoveConvertAfterGather::MoveConvertAfterGather() {
+    using namespace ov::pass::pattern;
+
+    // f16 compressed LLM word embedding pattern
+    // Const -(f16)-> Convert -(f32)-> Gather -(f32)->
```

Comment:
Could you clarify why Convert to f32 is kept in the model at all when we run with f16 inference precision? I'd expect that `ConvertPrecision` transformation which is called before `CommonOptimizations` pass (which triggers constant folding internally) replaces that with Convert(f16->f16) and then just eliminates it as NOP

Reply:
The `ConvertPrecision` pass does not convert precision to f16 for some precision sensitive operations, such as division with small number and so on. This is handled by the `MarkDivWithEps` inside `MarkSugraphsToKeepInMixedPrecision` as shown in the below codes.

https://github.com/openvinotoolkit/openvino/blob/a04d59d7792676fa60019e905c40bd4f5aebd62d/src/common/transformations/src/transformations/convert_precision.cpp#L394-L400

Qwen model has a pattern `Const(f16)` -- `Convert(f32)` -- `Gather` -- `Power`, and the last `Power` prevents precision  conversion to f16 for all layers that are connected to it. Next,`ConstantFolding` merges the two layers `Const(f16)` -- `Convert(f32)` into `Const(f32)`, and it results in larger memory usage. This PR prevents this situation by moving `Convert(f32)` after `Gather`.

Reply:
I see. @pavel-esir Is this expected behavior for precision sensitive `ConvertPrecision`? I think gather shouldn't be a part of sensitive subgraph, so it should be `Const(f16) -- Gather (f16) -- Convert (f32) -- Power (f32)` in the end, isn't it?

Reply:
If gather is connected to precision sensitive nodes it also becomes precision sensitive. On that time i did that intentionally, but at that time it didn't seem like it would cause memory problems.

It's worth considering to remove Gather from the list of ops through which precision sensitiveness can propagate but for now this solution is also good